### DISCUSSION TITLE
Ignore Websocket Upgrade Requests

### DISCUSSION THREAD

**Discussion by Argannor** [Kludex/uvicorn#1659] [Ideas] [answered]
Hi!
According to the HTTP 1.1 Specification a server may refuse an upgrade request by ignoring it and responding with a normal HTTP response instead (see [RFC7230 Section 6.7](https://datatracker.ietf.org/doc/html/rfc7230#section-6.7)). For me this behavior would be useful when running the application behind a proprietary reverse proxy that tries to upgrade to websocket (which I don't want).

Currently uvicorn responds with `No supported WebSocket library detected. Please use 'pip install uvicorn[standard]', or install 'websockets' or 'wsproto' manually.` and status code 400 or proceeds to upgrade to upgrade if a library is found.

My proposal is to add a configuration option "--ws-ignore-upgrade" and make the http implementations (http11, httptools) ignore the upgrade if the option is set. I'd like to contribute that change as well, if you agree with this request.
I'd also introduce a new method to both protocol implementations `should_upgrade(self, event) -> bool`, so the behaviour could be overwritten by someone needing more complex logic to decide whether to upgrade or not.

**Comment by iudeen** [comment]
Good idea

**Comment by Argannor** [comment]
Create a PR #1661 and included a solution for #1501 as well, since it's the same code location and it would conflict with #1642

**Comment by Kludex** [comment, marked as answer]
This is already implemented. :+1:
